src/main/frontend/handler/editor.cljs

```diff
@@ -2037,182 +1896,93 @@
-                  (outliner-core/save-node editing-block)))
-            _ (outliner-core/insert-nodes metadata-replaced-blocks target-block sibling?)
-            _ (when (and delete-editing-block? editing-block)
-                (when-let [id (:db/id editing-block)]
-                  (outliner-core/delete-node (outliner-core/block (db/pull id)) true)))]
-        (last metadata-replaced-blocks)))))
-
-(defn- tree->vec-tree
-  "tree:
-  [
-  {
-    :content 'this is a block',
-    :properties {\"key\" \"value\" \"key2\" \"value2\"},
-    :children [
-      { :content 'this is child block' }
-    ]
-  },
-  {
-    :content 'this is sibling block'
-  }
-  ]"
-  [tree]
-  (into []
-        (mapcat
-         (fn [e]
-           (let [e* (select-keys e [:content :properties])
-                 children (:children e)]
-             (if (seq children)
-               [e* (tree->vec-tree (:children e))]
-               [e*])))
-         tree)))
-
-(defn- vec-tree->vec-block-tree
-  [tree format]
-  (let [loc (zip/vector-zip tree)]
-    (loop [loc loc]
-      (if (zip/end? loc)
-        (zip/root loc)
-        (let [node (zip/node loc)]
-          (if (vector? node)
-            (recur (zip/next loc))
-            (let [content (:content node)
-                  props (into [] (:properties node))
-                  content* (str (if (= :markdown format) "- " "* ")
-                                (property/insert-properties format content props))
-                  ast (mldoc/->edn content* (mldoc/default-config format))
-                  blocks (block/extract-blocks ast content* true format)
-                  fst-block (first blocks)]
-              (assert fst-block "fst-block shouldn't be nil")
-              (recur (zip/next (zip/replace loc fst-block))))))))))
-
-(defn paste-block-tree-after-target
-  [target-block-id sibling? tree format]
-  (let [vec-tree (tree->vec-tree tree)
-        block-tree (vec-tree->vec-block-tree vec-tree format)
-        target-block (db/pull target-block-id)
-        page-block (if (:block/name target-block) target-block
-                       (db/entity (:db/id (:block/page (db/pull target-block-id)))))
-        ;; sibling? = false, when target-block is a page-block
-        sibling? (if (= target-block-id (:db/id page-block))
+                   (db/has-children? (:block/uuid target-block))
                    false
-                   sibling?)]
-    (paste-block-vec-tree-at-target
-     block-tree []
-     {:get-pos-fn #(get-block-tree-insert-pos-after-target target-block-id sibling?)
-      :page-block page-block})))
+
+                   :else
+                   true)]
+    (outliner-tx/transact!
+      {:outliner-op :insert-blocks}
+      (when editing-block
+        (outliner-core/save-block! editing-block))
+      (when target-block
+        (let [format (or (:block/format target-block) (state/get-preferred-format))
+              blocks (map (fn [block]
```

Comment:
Would recommend not clobbering existing names in the same scope. This combined with long fns makes it harder for others to understand and easier for us to make mistakes. If you can't think of a good name, you an always append a suffix e.g. `blocks_` or `blocks'`